### PR TITLE
Feature/api/announcement-database

### LINKED ISSUE #687: Funcionalidade de Banco de Editais
## Feature Description

Do mesmo molde do Banco de Ideias o banco de editais será possível visualizar os variados editais que são publicados por meio de outras instituições. Queremos dar a oportunidade para os nossos usuários fiquem sabendo quais editais vigentes eles podem consultar e participar.

Essa funcionalidade terá:

**1 - LISTA DE EDITAIS**
- Listar todos os editais cadastrados em formato de card (similar ao banco de ideias)
- O conteúdo dessa lista será:
   - Organização: 
   - Número do Edital: 
   - Título: 
   - Descritivo: 
   - Público-Alvo: 
   - Palavras-chave:
   - Recursos financeiros: 
   - Período de inscrição:
   - Observação:
   - Link para a Página do edital: (abrir em uma nova janela)

- Somente serão publicos os registros com status PUBLISHED
- Mesmo o edital com data expirada continua visível nessa lista 

**2 - REGISTRAR UM EDITAL**
- Qualquer usuário cadastrado poderá registrar um edital.
- Os campos serão:
   - Organização (Dropdown): 
   - Número do Edital: (campo texto)
   - Título: (campo texto)
   - Descritivo: (textarea)
   - Público-Alvo: (Lista de publico-alvo das taxonomias existentes)
   - Palavras-chave: (Lista de palavras-chave das taxonomias existentes)
   - Recursos financeiros: (Opcional) - campo de valor monetário
   - Período de inscrição: (data de inicio e data de fim)
   - Observação: (textarea)
   - Link para a Página do edital: (URL)

- Ao cadastrar o registro terá um status de pendente 
- Somente o ADMIN muda o status para PUBLISHED
- O ADMIN poderá cancelar o registro.

---------------

## Acceptance criteria

* 

## Implementation Brief

*

### PR DESCRIPTION
**Pull request by alexandreadames**
## Summary

Resolves #687 

### Rotas

| Description | Method | Route | Body/Query | Auth |
| --- | --- | --- | --- | --- |
Lists all Announcements. If user is ADMIN lists all status, otherwise only published.|GET| announcements | Query Params for filtering: title, description, keywords, targetAudiences | No | 
|Shows an announcement.If user is ADMIN or Owner gets all status, otherwise only published. |GET| announcements/:id | N/A | No |
|Any user with complete profile can creates an announcement |POST| announcements | institution_id (number), announcement_number(string), title (string),description (string), targetAudiences (term array id or slug),  keywords (term array id or slug), financial_resources (number), start_date(string), end_date (string), comment(string), url (string) | Yes |
|Announcement owner user can update it |PUT| announcements/:id | body POST | Yes Permission: UPDATE_ANNOUNCEMENT |
|ADMIN can update announcement status |PUT| announcements/:id/update-status | status='pending','published' | Yes role: ADMIN |
|Announcement owner user can delete it |DELETE| announcements/:id | N/A | Yes Permission: DELETE_ANNOUNCEMENT|

## QA Steps

Run `adonis test --files test/functional/announcement.spec.js`

## Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code passes the linting.
- [x] My code has proper inline documentation.
- [x] I have manually tested this PR.
- [x] Documentation with ApiDOC

### CODE REVIEW

**Review by NaylsonFerreira — DISMISSED** [review]
LGTM

**Review by NaylsonFerreira — DISMISSED** [review]

**Comment by mateus4k** [comment]
Please resolve conflicts

File: packages/api/test/functional/announcement.spec.js [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+			title: 'First 2021 Announcement',
+			description: 'First 2021 Announcement description',
+			keywords: keywordTermsIds,
+			targetAudiences: targetAudienceTermsIds,
+			financial_resources: 1000,
+			comment: 'single comment',
+			start_date: '2021-01-01',
+			end_date: '2021-01-31',
+			url: 'http://first2021annoucement.com',
+		})
```
Comment (mateus4k) [inline packages/api/test/functional/announcement.spec.js:151]: Vc pode substituir isso por um `const announcement = await Factory.model('App/Models/Announcement').make();`
```suggestion
		.send(announcement.toJSON())
```

File: packages/api/test/functional/announcement.spec.js [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+		.end();
+
+	const announcementCreated = await Announcement.findOrFail(response.body.id);
+	await announcementCreated.loadMany(['institution', 'terms']);
+
+	response.body.status = announcementStatuses.PENDING;
```
Comment (mateus4k) [inline packages/api/test/functional/announcement.spec.js:157]: Pq isso? 🤔🤔🤔

File: packages/api/test/functional/announcement.spec.js [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+		.loginVia(user, 'jwt')
+		.end();
+
+	response.assertStatus(403);
+	response.assertJSONSubset(
+		errorPayload(errors.UNAUTHORIZED_ACCESS, antl('error.permission.unauthorizedAccess')),
+	);
+});
+
+test('DELETE /announcements/:id deletes an announcements', async ({ client, assert }) => {
```
Comment (mateus4k) [inline packages/api/test/functional/announcement.spec.js:330]:
```suggestion
test('DELETE /announcements/:id deletes an announcement', async ({ client, assert }) => {
```

**Review by mateus4k — CHANGES_REQUESTED** [review]

File: packages/api/test/functional/announcement.spec.js (hunk 2 above)
Comment (alexandreadames) [reply packages/api/test/functional/announcement.spec.js:157]: Rapaz é porque quando eu busco o announcementCreated não vem o status

File: packages/api/app/Models/Announcement.js [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+		if (filters.title) {
+			query.where('title', 'LIKE', `%${filters.title}%`);
+		}
+		if (filters.description) {
+			query.where('description', 'LIKE', `%${filters.description}%`);
+		}
+
+		if (filters.keywords) {
+			const keywordsList = filters.keywords ? filters.keywords.split(',') : [];
+			if (keywordsList && keywordsList.length) {
```
Comment (NaylsonFerreira) [inline packages/api/app/Models/Announcement.js:56]:
```suggestion
			const keywordsList = filters?.keywords.split(',') || [];
			if (keywordsList?.length) {
```

File: packages/api/app/Models/Announcement.js [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+					builder.whereIn('id', keywordsList);
+				});
+			}
+		}
+
+		if (filters.targetAudiences) {
+			const targetAudiencesList = filters.targetAudiences
+				? filters.targetAudiences.split(',')
+				: [];
+			if (targetAudiencesList && targetAudiencesList.length) {
```
Comment (NaylsonFerreira) [inline packages/api/app/Models/Announcement.js:67]:
```suggestion
			const targetAudiencesList = filters?.targetAudiences.split(',') || [];
			if (targetAudiencesList?.length) {
```

File: packages/api/test/functional/announcement.spec.js [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+	const institution = await Factory.model('App/Models/Institution').create();
+	const publishedAnnouncements = await Factory.model('App/Models/Announcement').createMany(5);
+	await Promise.all(
+		publishedAnnouncements.map(async (announcement) => {
+			await announcement.user().associate(user);
+			await announcement.institution().associate(institution);
+			announcement.status = announcementStatuses.PUBLISHED;
+			await announcement.save();
+		}),
+	);
```
Comment (NaylsonFerreira) [inline packages/api/test/functional/announcement.spec.js:27]:
```suggestion
	await Promise.all(
		publishedAnnouncements.map((announcement) => {
			announcement.status = announcementStatuses.PUBLISHED;
			return [
				announcement.user().associate(user),
				announcement.institution().associate(institution),
				announcement.save(),
			];
		}),
	);
```

**Review by NaylsonFerreira — APPROVED** [review]

**Review by NaylsonFerreira — APPROVED** [review]
Boas Festas!!

**Review by NaylsonFerreira — APPROVED** [review]

**Review by mateus4k — APPROVED** [review]

**Pull request merged**